cloud/app/server/server.js

```diff
@@ -835,6 +839,57 @@ class _robotAgent extends Capability {
     });
   }
 
+  /** Stop cloud docker containers for capabilities that haven't been used in a
+    while. */
+  async stopUnusedContainers() {
+    if (process.env.NODOCKER) {
+      log.info('NODOCKER: not stopping unused docker containers');
+      return;
+    }
+    log.info('Checking for unused containers');
+
+    // flattens to list of container names, removing leadning @ and replacing / with .
+    const capabilitiesInUse = this.getCapabilitiesInUse();
+
+    log.info('Capabilities in use', capabilitiesInUse);
+    const capabilityContainers = await docker.listCapabilityContainers();
+    log.info('Running capability containers', capabilityContainers.map(c => c.Names));
+    _.forEach(capabilityContainers, async (container) => {
+      const containerName = container.Names[0].slice(1); // remove leading /
+      log.debug('stopUnusedContainers: checking', containerName);
+      // if no device is using this capability, stop it
+      const [scope, name, ...versionParts] = containerName.split('.');
+      const version = versionParts.join('.');
+      if (capabilitiesInUse[`@${scope}`]?.[name]?.[version]){
+        log.debug('stopUnusedContainers: still in use, not stopping', containerName);
+      } else {
+        // split name and version from containerName, e.g.
+        // transitive-robotics.terminal.0.1.5
+        log.info('Stopping', containerName, ' not in use by any device');
+        await docker.stop({name:`${scope}.${name}`, version});
+      }
+    });
+  }
+
+  /** Get last capabilities used by devices which have sent a heartbeat recently */
+  getCapabilitiesInUse() {
+    const capabilities = {};
+    _.forEach(this.data.get(), (orgData, orgId) => {
+      const devicesStatus = this.getStatus(orgId);
+      _.forEach(devicesStatus, (status, deviceId) => {
+        const heartbeat = new Date(status.heartbeat || 0).getTime();
+        // If device's heartbeat is old, don't consider its capabilities as in use
+        if (heartbeat < ( Date.now() - PRESERVE_CAPS_DOCKER_CONTAINERS_THRESHOLD)) return;
+          forMatchIterator(status, ['runningPackages', '+scope', '+capName', '+version'],
```

Comment:
wrong and very misleading indentation; on first sight this appear to be the body of the `if`.